### PR TITLE
Introduce trusted artifacts (take 2)

### PR DESCRIPTION
**Pull request by zregvart**
Adds the use of trusted artifacts[1] to the Tasks and Pipelines. The `prefetch-dependencies` Task is no longer skipped based on the `hermetic` parameter -- it needs to run always to produce the trusted artifacts as the build* tasks depend on these, so if it were skipped the build* tasks would be skipped as well.

The major change is that the `git-clone` and `prefetch-dependencies` Tasks do not leave files, apart from the trusted artifact, on the workspace (PVC). This increases the security as manipulation of the source files from a Task running in parallel is prevented.

Tasks now mount a local emptyDir-based volume in `/var/workdir` and operate within that. For example the sources are placed by the `use-trusted-artifact` steps in `/var/workdir/source` and if the Task depends on prefetched dependencies in `/var/workdir/cachi2`.

This moves the storage pressure from PVC to local storage, which can also have its own side-effects. For example the `clair-scan` Task, which expands the image, takes a large amount of disk space locally. Worth considering the benefits/risks of that.

Ref: https://issues.redhat.com/browse/EC-251

[1] https://github.com/redhat-appstudio/build-trusted-artifacts

### CODE REVIEW

**Comment by zregvart** [comment]
Another note, the PAC Tasks in `.tekton` have not been updated in this PR (they were in #712).

**Comment by zregvart** [comment]
Note for reviewers: notice the use of isolated workspaces (https://github.com/redhat-appstudio/build-definitions/pull/791/commits/3680cd006e6b229614a04d4f8e3fc42361e1d2b9)

**Review by chmeliik — COMMENTED** [review]
So far I've properly checked the first 4 commits and skimmed through the rest

File: task/git-clone/0.2/MIGRATION.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Migration from 0.1 to 0.2
+
+- The parameter `deleteExisting` used by `git-clone` task was removed. As the
+  source is no longer maintained in the workspace, all executions of the
+  `git-clone` Task will be clean fetches.
+- The parameter `subdirectory` used by `git-clone` task was removed.
```
[the remark this answers is not in the document]
Comment (zregvart) [reply task/git-clone/0.2/MIGRATION.md:6]: git-clone no longer leaves files in the workspace and always clones to a emptyDir volume, so those do not make sense any more

File: task/prefetch-dependencies/0.2/prefetch-dependencies.yaml [hunk 2]
```diff
@@ -29,6 +50,9 @@ spec:
     - name: HERMETIC
       value: $(params.hermetic)
     script: |
+      # create this directory so that the create-trusted-artifact step has
+      # something to work with, even though it might end up being empty
+      mkdir -p /var/workdir/cachi2
```
[the remark this answers is not in the document]
Comment (zregvart) [reply task/prefetch-dependencies/0.2/prefetch-dependencies.yaml:57]: We can fix that in https://github.com/redhat-appstudio/build-trusted-artifacts, but just haven't yet

File: task/buildah/0.2/MIGRATION.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  `CACHI2_ARTIFACT` from the results of the `prefetch-dependencies` Task
+
+```diff
+ - name: build-container
+   params:
++  - name: SOURCE_ARTIFACT
++    value: $(tasks.prefetch-dependencies.results.SOURCE_ARTIFACT)
++  - name: CACHI2_ARTIFACT
++    value: $(tasks.prefetch-dependencies.results.CACHI2_ARTIFACT)
+```
```
[the remark this answers is not in the document]
Comment (zregvart) [reply task/buildah/0.2/MIGRATION.md:34]: Someone could have customized the pipeline and removed the `prefetch-dependencies` Task altogether, AFAIK this is allowed

File: task/buildah/0.2/README.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+|CACHI2_ARTIFACT|Trusted artifact containing the prefetched dependencies||false|
+|DOCKERFILE|Path to the Dockerfile to build.|./Dockerfile|false|
+|CONTEXT|Path to the directory to use as context.|.|false|
+|TLSVERIFY|Verify the TLS on the registry endpoint (for push/pull to a non-TLS registry)|true|false|
+|HERMETIC|Determines if build will be executed without network access.|false|false|
+|PREFETCH_INPUT|In case it is not empty, the prefetched content should be made available to the build.|""|false|
+|IMAGE_EXPIRES_AFTER|Delete image tag after specified time. Empty means to keep the image tag. Time values could be something like 1h, 2d, 3w for hours, days, and weeks, respectively.|""|false|
+|YUM_REPOS_D_SRC|Path in the git repository in which yum repository files are stored|repos.d|false|
+|YUM_REPOS_D_FETCHED|Path in source workspace where dynamically-fetched repos are present|fetched.repos.d|false|
+|YUM_REPOS_D_TARGET|Target path on the container in which yum repository files should be made available|/etc/yum.repos.d|false|
```
[the remark this answers is not in the document]
Comment (zregvart) [reply task/buildah/0.2/README.md:22]: It was based on it a while back, so changes could have been introduced, will double check

[the remark this answers is not in the document]
Comment (chmeliik) [reply task/buildah/0.2/README.md:22]: Right - #771 had to be dropped, so some of the changes need to be un-done :sweat_smile:

File: task/buildah/0.2/MIGRATION.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply task/buildah/0.2/MIGRATION.md:34]: Hmm, that is true. How about wording it like "if your pipeline includes a prefetch-dependencies task ..."

**Comment by chmeliik** [comment]
> In general, the number of required migrations is similarly concerning as it was in #771
> 
> Although here it's mostly the same changes for all the migrations, so maybe it's not _that_ bad? And it shouldn't be more than 4 or 5 tasks to migrate per pipeline (clone, prefetch, build, build-source, snyk)

Idea - how hard would it be to write a migration script that makes the relevant trusted-artifact changes in the PipelineRun yamls? We could link that script from the migration docs

File: task/git-clone/0.2/git-clone.yaml [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+      value: $(params.fetchTags)
+    - name: WORKSPACE_SSH_DIRECTORY_BOUND
+      value: $(workspaces.ssh-directory.bound)
+    - name: WORKSPACE_SSH_DIRECTORY_PATH
+      value: $(workspaces.ssh-directory.path)
+    - name: WORKSPACE_BASIC_AUTH_DIRECTORY_BOUND
+      value: $(workspaces.basic-auth.bound)
+    - name: WORKSPACE_BASIC_AUTH_DIRECTORY_PATH
+      value: $(workspaces.basic-auth.path)
+    image: registry.redhat.io/openshift-pipelines/pipelines-git-init-rhel8:v1.8.2-8@sha256:a538c423e7a11aae6ae582a411fdb090936458075f99af4ce5add038bb6983e8
```
[the remark this answers is not in the document]
Comment (zregvart) [reply task/git-clone/0.2/git-clone.yaml]: I've just rebased with the intent to make the changes between 0.1 and 0.2 minimal. Though it might be a good opportunity to remove those deprecated parameters in 0.2 (so we don't need a 0.3). Happy to include that here as well.

File: task/prefetch-dependencies/0.2/prefetch-dependencies.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (zregvart) [reply task/prefetch-dependencies/0.2/prefetch-dependencies.yaml:57]: [This](https://github.com/redhat-appstudio/build-trusted-artifacts/pull/21) should fix the issue over there, will update here once it's released

**Comment by zregvart** [comment]
> Idea - how hard would it be to write a migration script that makes the relevant trusted-artifact changes in the PipelineRun yamls? We could link that script from the migration docs

I don't think that would be that hard to do, created https://issues.redhat.com/browse/EC-418 to followup on that.

File: task/buildah/0.2/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (zregvart) [reply task/buildah/0.2/README.md:24]: Happy to include that here, but it might be a better idea to have it fixed in main and rebase here on top of that to get the README updated for 0.1 version as well...

File: task/git-clone/0.2/git-clone.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (zregvart) [reply task/git-clone/0.2/git-clone.yaml:120]: In lieu of a migration tool, perhaps that decision could be reevaluated? Eventually we will remove those parameters, best scar the users just once perhaps?

The latest change removes the `gitInitImage` and `userHome` Task parameters.

File: task/buildah/0.2/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply task/buildah/0.2/README.md:24]: Ah, I didn't notice the param was missing in the README in main. I mainly wanted to point out that the param was missing from the task yaml itself (fixed now)

[the remark this answers is not in the document]
Comment (zregvart) [reply task/buildah/0.2/README.md:24]: README now includes the trusted artifacts and the removal of `gitInitImage` and `userHome` parameters...

File: task/prefetch-dependencies/0.2/prefetch-dependencies.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (zregvart) [reply task/prefetch-dependencies/0.2/prefetch-dependencies.yaml]: Updated the Task to remove the `mkdir` and updated to the new version of TA

File: task/git-clone/0.2/git-clone.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply task/git-clone/0.2/git-clone.yaml:120]: There are plans to look into migration tools ([STONEBLD-2147](https://issues.redhat.com/browse/STONEBLD-2147)), until then it was decided that removing deprecated params is not worth it (except for those that are unsafe, like the `*image` params)

File: task/buildah/0.2/MIGRATION.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  results of the `prefetch-dependencies` Task instead
+
+```diff
+ - name: build-container
+   params:
++  - name: SOURCE_ARTIFACT
++    value: $(tasks.prefetch-dependencies.results.SOURCE_ARTIFACT)
++  - name: CACHI2_ARTIFACT
++    value: $(tasks.prefetch-dependencies.results.CACHI2_ARTIFACT)
+```
```
[the remark this answers is not in the document]
Comment (zregvart) [reply task/buildah/0.2/MIGRATION.md:35]: Easy to replace the order here...

File: task/git-clone/0.2/git-clone.yaml [hunk 7]
```diff
@@ -57,18 +57,6 @@ spec:
...
-    # per https://kubernetes.io/docs/concepts/containers/images/#imagepullpolicy-defaulting
-    # the cluster will set imagePullPolicy to IfNotPresent
-    description: The image providing the git-init binary that this Task runs.
-    name: gitInitImage
-    type: string
-  - default: /tekton/home
-    description: |
-      Absolute path to the user's home directory. Set this explicitly if you are running the image as a non-root user or have overridden
-      the gitInitImage param with an image containing custom user configuration.
-    name: userHome
```
[the remark this answers is not in the document]
Comment (zregvart) [reply task/git-clone/0.2/git-clone.yaml]: Perhaps, my thinking is that we control this via the image used and the securityContext/runAsUser and without changing the image there is no use for `userHome` parameter...

File: task/buildah/0.2/MIGRATION.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply task/buildah/0.2/MIGRATION.md:35]: Maybe remove the "maintaining the hermetic build" as well

File: task/git-clone/0.2/git-clone.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply task/git-clone/0.2/git-clone.yaml]: TBH I'm not sure if the user has the ability to override runAsUser or not. But if we eventually get rid of runAsUser (hardcoding root doesn't seem like a good idea long term), then we will want that param back

[the remark this answers is not in the document]
Comment (zregvart) [reply task/git-clone/0.2/git-clone.yaml]: The latest version includes the `userHome` parameter.

File: task/rpm-ostree/0.2/MIGRATION.md [hunk 8]
```diff
@@ -7,15 +7,27 @@ artifact, e.g. using the git-clone Task version 0.3 or newer.
...
 
 ```diff
  - name: rpm-ostree
    params:
 +  - name: SOURCE_ARTIFACT
-+    value: $(tasks.clone-repository.results.SOURCE_ARTIFACT) # clone-repository is the name of the git-clone Task in the Pipeline
++    value: $(tasks.prefetch-dependencies.results.SOURCE_ARTIFACT)
++  - name: CACHI2_ARTIFACT
++    value: $(tasks.prefetch-dependencies.results.CACHI2_ARTIFACT)
 ```
```
Comment (chmeliik) [inline task/rpm-ostree/0.2/MIGRATION.md:21]: The rpm-ostree task doesn't actually accept a CACHI2_ARTIFACT param

There's no dedicated pipeline that would use rpm-ostree as the build task (weird), but when I look at https://github.com/CentOS/centos-bootc/blob/main/.tekton/ostree-build.yaml, it doesn't use the prefetch task.

TLDR we should probably keep only the case where the pipeline doesn't include prefetch-dependencies

File: task/git-clone/0.2/README.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+|depth|Perform a shallow clone, fetching only the most recent N commits.|1|false|
+|sslVerify|Set the `http.sslVerify` global git config. Setting this to `false` is not advised unless you are sure that you trust your git remote.|true|false|
+|subdirectory|Subdirectory inside the `output` Workspace to clone the repo into.|""|false|
+|sparseCheckoutDirectories|Define the directory patterns to match or exclude when performing a sparse checkout.|""|false|
+|deleteExisting|Clean out the contents of the destination directory if it already exists before cloning.|true|false|
+|httpProxy|HTTP proxy server for non-SSL requests.|""|false|
+|httpsProxy|HTTPS proxy server for SSL requests.|""|false|
+|noProxy|Opt out of proxying HTTP/HTTPS requests.|""|false|
+|verbose|Log the commands that are executed during `git-clone`'s operation.|true|false|
+|userHome|Absolute path to the user's home directory. Set this explicitly if you are running the image as a non-root user or have overridden the gitInitImage param with an image containing custom user configuration. |/tekton/home|false|
```
Comment (chmeliik) [inline task/git-clone/0.2/README.md:21]: We should drop the mention of the `gitInitImage` param in the description of `userHome`

**Review by chmeliik — COMMENTED** [review]
Last two comments for the documentation, otherwise LGTM

**Review by chmeliik — APPROVED** [review]
LGTM :tada: 

Please re-update based on main just before merging, to avoid losing some image updates that may have happened in the 0.1 versions of tasks

**Comment by chmeliik** [comment]
Needs another rebase

Pls ping me afterwards so that we can get this merged before more conflicts appear in main

**Comment by zregvart** [comment]
I'm waiting on the results of the performance tests so I don't want to change a lot here till the work is done there, but I have a branch ready to push with the rebase.

**Comment by zregvart** [comment]
Performance tests are not very favorable, see [KONFLUX-1031](https://issues.redhat.com/browse/KONFLUX-1031) for details. I'm going to convert this to draft so we can discuss this further.

**Comment by lcarva** [comment]
We should probably close this. Related: https://github.com/redhat-appstudio/build-definitions/pull/913#issuecomment-2105275146
